ducktape/src/main/scala/io/github/arainko/ducktape/internal/macros/FallibleCoproductTransformations.scala

```diff
@@ -0,0 +1,84 @@
+package io.github.arainko.ducktape.internal.macros
+
+import io.github.arainko.ducktape.fallible.{ FallibleTransformer, Mode }
+import io.github.arainko.ducktape.function.FunctionArguments
+import io.github.arainko.ducktape.internal.modules.MaterializedConfiguration.*
+import io.github.arainko.ducktape.internal.modules.*
+
+import scala.deriving.*
+import scala.quoted.*
+
+private[ducktape] object FallibleCoproductTransformations {
+
+  def transform[F[+x]: Type, Source: Type, Dest: Type](
+    Source: Expr[Mirror.SumOf[Source]],
+    Dest: Expr[Mirror.SumOf[Dest]],
+    F: Expr[Mode[F]],
+    sourceValue: Expr[Source]
+  )(using Quotes): Expr[F[Dest]] = {
+    given Cases.Source = Cases.Source.fromMirror(Source)
+    given Cases.Dest = Cases.Dest.fromMirror(Dest)
+
+    val ifBranches = coproductBranches[F, Source, Dest](sourceValue, Cases.source.value, F)
+    ifStatement(ifBranches).asExprOf[F[Dest]]
+  }
+
+  private def coproductBranches[F[+x]: Type, Source: Type, Dest: Type](
+    sourceValue: Expr[Source],
+    sourceCases: List[Case],
+    F: Expr[Mode[F]]
+  )(using Quotes, Cases.Dest) = {
+    import quotes.reflect.*
+
+    sourceCases.map { source =>
+      source -> Cases.dest
+        .get(source.name)
+        .getOrElse(Failure.emit(Failure.NoChildMapping(source.name, summon[Type[Dest]])))
+    }.map { (source, dest) =>
+      val cond = IsInstanceOf(sourceValue, source.tpe)
+
+      (source.tpe -> dest.tpe) match {
+        case '[src] -> '[dest] =>
+          val value =
+            source.fallibleTransformerTo[F](dest).map {
```

Comment:
there's an optimization we can do here where we match on the summoned transformer and if it is a `FallibleTransformer.fallibleFromTotal` we can strip apart the total transformer with `LiftTransformation.liftTransformation` which will make it so that the body of the transformer is inlined instead of creating a transformer instance at runtime. Example of how to do it is here: https://github.com/arainko/ducktape/blob/1c50c0ef64a1acea6207553baf119471430351c2/ducktape/src/main/scala/io/github/arainko/ducktape/internal/macros/AccumulatingProductTransformations.scala#L34

Reply:
I think I've achieved what you had suggested. I had problems getting a test case to match `FallibleTransformer.fallibleFromTotal`. But I've added an optimization to use total transformers, if available. The cost for that optimization is a second implicit search. I'm not sure if that's worth it.